### DISCUSSION TITLE
Query carla parameters via network

### DISCUSSION THREAD

**Discussion by kuon** [falkTX/Carla#1422] [Q&A]
I can control carla fine with command like:

`oscsend localhost 22752 /Carla/10/set_volume f 0  `

But I'd like to be able to query the parameters for scripting.

If there is another approach (maybe some midi tool?) to query the parameter from the command line, that would work.

**Comment by falkTX** [comment]
This should be a discussion, not a issue ticket.
Lets continue there

**Reply by kuon** [reply]
Sorry, I didn't know there was now discussion on github.

**Reply by falkTX** [reply]
no worries, it is a brand new thing overall

**Comment by falkTX** [comment]
So you can use what carla-control does.
you need to have both a TCP and UDP OSC server running, see https://github.com/falkTX/Carla/blob/main/source/frontend/carla_host_control.py#L269

TCP is for sending critical messages, UDP is for peak meters and other stuff that is okay to drop packets for.

you can activate the whole remote control by sending a /register message which starts the whole thing.

there is no documentation for all this though, sorry.
I dont consider this remote control stuff to be stable enough yet.
for one, there is no way to query plugin list due to some bad core designs.

**Reply by kuon** [reply]
I would really like to have a way to query parameters from scripts. Without having to spawn an entire server.

The idea is to be able to display some parameters in my desktop environment. So I wanted something lightweight.

I guess I could write a small middleware server that talk with carla.

One other approach I was looking into is to use MIDI. Having a midi plugin that controls the parameters and this plugin would have a command line API to set and query parameters, the advantage of this approach is that I can wire things in carla. But I don't know were to start and if it can work.

**Reply by falkTX** [reply]
the most basic thing for this that I have sorta in the plans is an HTTP/REST API.
so you then could just "curl" the API and get a response back.

I tried this already using restbed as C++ framework for the HTTP handling, but have too many crashes.

I would prefer to go into this direction, since it is something I want to see added in carla in any case.

**Reply by kuon** [reply]
I guess a rest or websocket API would be the best. In the meantime, I just set the parameter with OSC and write my setting in a file (so if I change it in carla, it's out of sync, but it works for my use).

So basically, my command line tool `osc_remote set x y` will do OSC send to carla and write `x=y` in it's config file at `.config/osc_remote/params`. Then `osc_remote get x` just return the stored value.

It's brittle, but work for my use, until carla has a queryable endpoint.

**Comment by rbren** [comment]
This would be an amazing feature!

My end-goal is to be able to script loops/songs that draw from SFZ and other instruments, say in python or some home-grown DSL.

Of course I don't want to deal with all the low-level audio sampling and processing, so my hope is that I can script around Carla. Specifically I'm hoping for an API that lets me:
* load new instruments into the rack
* set up the patchbay topology
* send a stream of MIDI notes and other parameters
* get an audio stream back from Carla
* get a stream of other events (e.g. MIDI keypresses or knob turns) back from Carla

Does the HTTP API support these today? I'd be happy to dig through code and help document this if you can point me in the right direction.

**Comment by rbren** [comment]
It seems like there's _something_ running at `localhost:22752`, but I can't connect via curl or [websocat](https://github.com/vi/websocat) (I tried `websocat ws://127.0.0.1:22752/ws`). Everything just hangs, even though `carla-control` connected just fine.

[this line](https://github.com/falkTX/Carla/blob/25c61989de0cfe55e4ff62af002403c9acaa1bd6/source/frontend/carla_backend_qtweb.py#L41) seems to suggest that the port is actually hard-coded to `2228`, which doesn't seem to be listening.

**Comment by rbren** [comment]
OK I've managed to connect via OSC. I'm seeing some output in Carla's logs, but haven't managed to send a valid message yet 😓 

Best I've gotten is this python code:
```python
# Import needed modules from osc4py3
from osc4py3.as_eventloop import *
from osc4py3 import oscbuildparse
import time

# Start the system.
osc_startup()

# Make client channels to send packets.
osc_udp_client("127.0.0.1", 22752, "aclientname")

# Build a simple message and send it.
msg = oscbuildparse.OSCMessage("/Carla/1/set_volume", ",sif", ["text", 672, 0])
osc_send(msg, "aclientname")

# Periodically call osc4py3 processing method in your event loop.
finished = False
while not finished:
    # You can send OSC messages from your event loop too…
    # …
    print("processing")
    osc_process()
    print("processed")
    # …

# Properly close the system.
osc_terminate()
```

Carla logs say
```
invalid plugin id '1', probably has been removed (path: '/Carla/1/set_volume')
```

**Comment by rbren** [comment]
I should also note:

Valid OSC endpoints seem to be [here](https://github.com/falkTX/Carla/blob/15f7f584f651a2d590e33171f7f0d7f1110c0288/source/backend/engine/CarlaEngineOscHandlers.cpp#L168)

Which is a subset of [what's listed here](https://kx.studio/ns/dev-docs/CarlaBackend/group__CarlaHostAPI.html) (I'd found that mentioned on a previous issue).

There's a much larger set of REST endpoints listed [here](https://github.com/falkTX/Carla/blob/25c61989de0cfe55e4ff62af002403c9acaa1bd6/source/rest/rest-server.cpp#L330) (C++) and [here](https://github.com/falkTX/Carla/blob/25c61989de0cfe55e4ff62af002403c9acaa1bd6/source/frontend/carla_backend_qtweb.py#L447) (python), but I can't figure out how to connect to that server.

**Comment by rbren** [comment]
@falkTX would love your advice on how to proceed--I'm a bit stuck at this point.
